### PR TITLE
Suggest using precise capturing for hidden type that captures region

### PR DESCRIPTION
**Pull request by compiler-errors**
Adjusts the "add `+ '_`" suggestion for opaques to instead suggest adding or reusing the `+ use<>` in the opaque.

r? oli-obk or please re-roll if you're busy!

Tracking:

- https://github.com/rust-lang/rust/issues/123432

### CODE REVIEW

File: tests/ui/impl-trait/precise-capturing/forgot-to-capture-lifetime.stderr [hunk 1]
```diff
@@ -16,10 +16,10 @@ LL | fn lifetime_in_hidden<'a>(x: &'a ()) -> impl Sized + use<> { x }
    |                       |                 opaque type defined here
    |                       hidden type `&'a ()` captures the lifetime `'a` as defined here
    |
-help: to declare that `impl Sized` captures `'a`, you can add an explicit `'a` lifetime bound
+help: add `'a` to the `use<...>` bound to explicitly capture it
    |
-LL | fn lifetime_in_hidden<'a>(x: &'a ()) -> impl Sized + use<> + 'a { x }
-   |                                                            ++++
+LL | fn lifetime_in_hidden<'a>(x: &'a ()) -> impl Sized + use<'a> { x }
```
Comment (compiler-errors) [inline tests/ui/impl-trait/precise-capturing/forgot-to-capture-lifetime.stderr:21]: ✨ yay

File: compiler/rustc_infer/src/infer/error_reporting/region.rs [hunk 2]
```diff
@@ -1201,17 +1202,21 @@ pub fn unexpected_hidden_region_diagnostic<'a, 'tcx>(
...
-                    tcx,
-                    &mut err,
-                    fn_returns,
-                    hidden_region.to_string(),
-                    None,
-                    format!("captures `{hidden_region}`"),
-                    None,
-                    Some(reg_info.def_id),
-                )
+                if infcx.tcx.features().precise_capturing {
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_infer/src/infer/error_reporting/region.rs:1205]: Yeah, that could be done, though I think we plan on stabilizing precise capturing **soon**, so not totally certain about the value. Also, having things gated on "is this a nightly compiler" adds one more weird case to test 😅

**Review by oli-obk — APPROVED** [review]
r=me with or without nit if you think it's overkill

**Comment by compiler-errors** [comment]
Actually I think I'll just open a follow-up that at least leaves a `note` explaining what needs to be done here to fix the code.

@bors r=oli-obk rollup

File: compiler/rustc_infer/src/infer/error_reporting/region.rs [hunk 3]
```diff
@@ -1257,3 +1262,98 @@ pub fn unexpected_hidden_region_diagnostic<'a, 'tcx>(
...
+                }
+
+                match param.kind {
+                    ty::GenericParamDefKind::Lifetime => {
+                        captured_lifetimes.insert(param.name);
+                    }
+                    ty::GenericParamDefKind::Type { synthetic: true, .. } => {
+                        // FIXME: We can't provide a good suggestion for
+                        // `use<...>` if we have an APIT. Bail for now.
+                        return;
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_infer/src/infer/error_reporting/region.rs:1322]: better yet: #127664

**Pull request merged**